ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/handler/TestHandler.java

```diff
@@ -0,0 +1,149 @@
+    addBaseFile(t, p, 25L, 25);
+
+    burnThroughTransactions(t.getDbName(), t.getTableName(), 25);
+
+    CompactionRequest rqst = new CompactionRequest(t.getDbName(), t.getTableName(), CompactionType.MAJOR);
+    rqst.setPartitionname("ds=today");
+    compactInTxn(rqst);
+
+    Handler handler = new CompactionHandler(conf, txnHandler, false);
+
+    // Fetch the compaction request using the handler
+    List<CleaningRequest> cleaningRequests = handler.findReadyToClean();
+    Assert.assertEquals(1, cleaningRequests.size());
+    CleaningRequest cr = cleaningRequests.get(0);
+    Assert.assertEquals(t.getDbName(), cr.getDbName());
+    Assert.assertEquals(t.getTableName(), cr.getTableName());
+    Assert.assertEquals("ds=today", cr.getPartitionName());
+    Assert.assertEquals(CleaningRequest.RequestType.COMPACTION, cr.getType());
+
+    // Check whether appropriate handler utility methods are called exactly once in a successful compaction scenario.
+    Handler mockedHandler = Mockito.spy(handler);
+    AtomicBoolean stop = new AtomicBoolean(true);
+    Cleaner cleaner = new Cleaner(Arrays.asList(mockedHandler));
+    cleaner.setConf(conf);
+    cleaner.init(stop);
+    cleaner.run();
+
+    Mockito.verify(mockedHandler, Mockito.times(1)).findReadyToClean();
+    Mockito.verify(mockedHandler, Mockito.times(1)).beforeExecutingCleaningRequest(any(CleaningRequest.class));
+    Mockito.verify(mockedHandler, Mockito.times(1)).afterExecutingCleaningRequest(any(CleaningRequest.class), any(List.class));
+  }
+
+  @Test
+  public void testCompactionHandlerForFailureCompaction() throws Exception {
+    Table t = newTable("default", "handler_failure_table", true);
+    Partition p = newPartition(t, "today");
+    addBaseFile(t, p, 20L, 20);
+    addDeltaFile(t, p, 21L, 22L, 2);
+    addDeltaFile(t, p, 23L, 24L, 2);
+    addBaseFile(t, p, 25L, 25);
+
+    burnThroughTransactions(t.getDbName(), t.getTableName(), 25);
+
+    CompactionRequest rqst = new CompactionRequest(t.getDbName(), t.getTableName(), CompactionType.MAJOR);
+    rqst.setPartitionname("ds=today");
+    compactInTxn(rqst);
+
+    // Check whether appropriate handler utility methods are called exactly once in a failure compaction scenario.
+    TxnStore mockedTxnHandler = Mockito.spy(txnHandler);
+    doThrow(new RuntimeException()).when(mockedTxnHandler).markCleaned(any());
+    Handler mockedHandler = Mockito.spy(new CompactionHandler(conf, mockedTxnHandler, false));
+    AtomicBoolean stop = new AtomicBoolean(true);
+    Cleaner cleaner = new Cleaner(Arrays.asList(mockedHandler));
+    cleaner.setConf(conf);
+    cleaner.init(stop);
+    cleaner.run();
+
+    Mockito.verify(mockedHandler, Mockito.times(1)).findReadyToClean();
+    Mockito.verify(mockedHandler, Mockito.times(1)).beforeExecutingCleaningRequest(any(CleaningRequest.class));
+    Mockito.verify(mockedHandler, Mockito.times(1)).failureExecutingCleaningRequest(any(CleaningRequest.class), any(Exception.class));
```

Comment:
You could also assert that afterExecutingCleaningRequest was not invoked.

Reply:
Added assert. Done.